dbms/src/AggregateFunctions/AggregateFunctionEntropy.h

```diff
@@ -0,0 +1,161 @@
+#pragma once
+
+#include <AggregateFunctions/FactoryHelpers.h>
+
+/// These must be exposed in header for the purpose of dynamic compilation.
+#include <AggregateFunctions/QuantileReservoirSampler.h>
+#include <AggregateFunctions/QuantileReservoirSamplerDeterministic.h>
+#include <AggregateFunctions/QuantileExact.h>
+#include <AggregateFunctions/QuantileExactWeighted.h>
+#include <AggregateFunctions/QuantileTiming.h>
+#include <AggregateFunctions/QuantileTDigest.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <AggregateFunctions/QuantilesCommon.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnDecimal.h>
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+
+#include <type_traits>
+
+#include <cmath>
+#include <AggregateFunctions/UniqVariadicHash.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+
+/** Calculates Shannon Entropy, using HashMap and computing empirical distribution function
+  */
+template <typename Value>
+struct EntropyData
+{
+    using Weight = UInt64;
+    using Map = HashMap <
+    Value, Weight,
+    HashCRC32<Value>,
+    HashTableGrower<4>,
+    HashTableAllocatorWithStackMemory<sizeof(std::pair<Value, Weight>) * (1 << 3)>
+    >;
+
+    Map map;
+
+    void add(const Value &x)
+    {
+        if (!isNaN(x))
+            ++map[x];
+    }
+
+    void add(const Value &x, const Weight &weight)
+    {
+        if (!isNaN(x))
+            map[x] += weight;
+    }
+
+    void merge(const EntropyData &rhs)
+    {
+        for (const auto &pair : rhs.map)
+            map[pair.first] += pair.second;
+    }
+
+    void serialize(WriteBuffer &buf) const
+    {
+        map.write(buf);
+    }
+
+    void deserialize(ReadBuffer &buf)
```

Comment:
To test serialization, we can write distributed query by select from `remote` table function with `remote('127.0.0.{1,2}')`. See examples in other tests. 

Note: To run this test on local machine, the server should be configured to accept remote connections (see `listen` in config.xml).